test/modules/CloudToDeviceMessageTester/Receiver.cs

```
@@ -0,0 +1,71 @@
+            : base(logger, iotHubConnectionString, deviceId, moduleId, transportType, testDuration, testResultReportingClient)
+        {
+        }
+
+        public override void Dispose() => this.deviceClient?.Dispose();
+
+        public async Task ReportTestResult(Message message)
+        {
+            string trackingId = message.Properties[TestConstants.Message.TrackingIdPropertyName];
+            string batchId = message.Properties[TestConstants.Message.BatchIdPropertyName];
+            string sequenceNumber = message.Properties[TestConstants.Message.SequenceNumberPropertyName];
+            var testResultReceived = new MessageTestResult(this.moduleId + ".receive", DateTime.UtcNow)
+            {
+                TrackingId = trackingId,
+                BatchId = batchId,
+                SequenceNumber = sequenceNumber
+            };
+            await ModuleUtil.ReportTestResultAsync(this.testResultReportingClient, this.logger, testResultReceived);
+        }
+
+        public override async Task InitAsync(CancellationTokenSource cts, DateTime testStartAt)
+        {
+            Microsoft.Azure.Devices.RegistryManager registryManager = null;
+            try
+            {
+                registryManager = Microsoft.Azure.Devices.RegistryManager.CreateFromConnectionString(this.iotHubConnectionString);
+                await registryManager.AddDeviceAsync(new Microsoft.Azure.Devices.Device(this.deviceId + "_leaf"), cts.Token);
+                this.deviceClient = DeviceClient.CreateFromConnectionString(this.iotHubConnectionString, this.deviceId, Microsoft.Azure.Devices.Client.TransportType.Amqp_Tcp_Only);
+                await this.deviceClient.OpenAsync();
+                while (!cts.IsCancellationRequested && this.IsTestTimeUp(testStartAt))
+                {
+                    this.logger.LogInformation("Ready to receive message");
+                    Message message = await this.deviceClient.ReceiveAsync();
```

what happens if there is not message sending to this module; will this line throw exception? e.g. timeout exception.

---

(as we discussed in person): It will. And then it will bubble up to program, where it is caught, logged, and test is ended. That will signify a problem with our setup, so failing fast is a good thing here.